### PR TITLE
Updating BMC ACPI power state.

### PR DESCRIPTION
**Pull request by purplenogg731**
## Description

Add ExitBootService event and S5 SMI handler for updating BMC ACPI power state.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

1. Power on SUT.
2. Check if SUT boot into Shell.
3. Change ACPI Power State to S3/D3.
  ipmitool -H 10.134.67.13 -U admin -P admin -I lanplus raw 0x6 0x6 0x83 0x83
  ipmitool -H 10.134.67.13 -U admin -P admin -I lanplus raw 0x6 0x7
   03* 03**
4. Boot into Windows.
5. Check if state has been changed to S0/D0.
  ipmitool -H 10.134.67.13 -U admin -P admin -I lanplus raw 0x6 0x7
   00* 00**
6. Shutdown SUT.
7. Check if state has been changed to S5/D3.
  ipmitool -H 10.134.67.13 -U admin -P admin -I lanplus raw 0x6 0x7
   05* 03**

> Note:
> *00h  S0 / G0   working
> *03h  S3        typically equates to “suspend-to-RAM”
> *05h  S5 / G2   soft off
> 
> **00h  D0
> **03h  D3

## Integration Instructions

N/A

### CODE REVIEW

**Comment by purplenogg731** [comment]
@microsoft-github-policy-service agree

File: IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
+/** @file
+    BmcAcpiPowerStateSmm.c
+
+    Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cfernald) [inline IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.c:4]:
```suggestion
    Copyright (c) Microsoft Corporation. All rights reserved.
    SPDX-License-Identifier: BSD-2-Clause-Patent
```

File: IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.inf [hunk 2]
```diff
@@ -0,0 +1,42 @@
+## @file BmcAcpiPowerStateSmm.inf
+#
+# INF description file for BmcAcpiPowerState SMM driver
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cfernald) [inline IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.inf:5]: Needs the SPDX tag

```suggestion
# Copyright (c) Microsoft Corporation. All rights reserved.
# SPDX-License-Identifier: BSD-2-Clause-Patent
```

File: IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.c [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  IN  UINT8  SystemPowerState,
+  IN  UINT8  DevicePowerState
+  )
+{
+  EFI_STATUS                         Status;
+  IPMI_SET_ACPI_POWER_STATE_REQUEST  SetAcpiPowerState;
+  UINT8                              CompletionCode;
+  UINT32                             DataSize;
+
+  SetAcpiPowerState.SystemPowerState.Bits.StateChange = 1;
```
Comment (cfernald) [inline IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.c:36]: Probably worth zeroing SetAcpiPowerState, event if all fields are set. e.g. ZeroMem(&SetAcpiPowerState, sizeof(SetAcpiPowerState);

Comment (purplenogg731) [reply IpmiFeaturePkg/BmcAcpiPowerState/BmcAcpiPowerStateSmm.c:39]: Added code to initialize SetAcpiPowerState in commit ba9c36e47bd98b1c9a33fcc9b64c018135455ac0. Thank you for your suggetion.

**Review by cfernald — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

**Pull request merged**